### PR TITLE
ninja_1_11: init at 1.11.1

### PR DESCRIPTION
**Pull request by emilazy**
Swift 5.8 doesn’t reliably build with Ninja 1.12, and apparently neither do old versions of ROCm, due to a build scheduling change breaking builds with undeclared dependencies. Upstream Swift have pinned 1.11 for now.

Reintroducing it isn’t a good long‐term solution, but seems like the easiest way to avoid shipping 24.11 with a broken Swift on Darwin. The patch for 32‐bit systems has not been restored, as Swift doesn’t support them anyway.

cc @azuwis @JohnRTitor. I think this is a better approach than `overrideAttrs`. I confirmed that `nix-update` can still update the package in this form.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ni/ninja/package.nix
Comment (paparodeo) [inline pkgs/by-name/ni/ninja/package.nix:27]:
```suggestion
    "latest" = fetchFromGitHub {
```

this one should be the latest version, we're not trying to pin it to any specific version.

Comment (emilazy) [reply pkgs/by-name/ni/ninja/package.nix]: Fair enough; pushed. Hopefully this is gone by the time 1.13 comes around either way.

**Comment by emilazy** [comment]
De jure Ninja maintainers don’t seem to be actively involved; cc @JohnRTitor @vcunat @NickCao @AndersonTorres who worked on this derivation since 2023.

File: pkgs/by-name/ni/ninja/package.nix
Comment (paparodeo) [inline pkgs/by-name/ni/ninja/package.nix:14]:
```suggestion
, use_1_11 ? false
```

nit. a bool rather than a string would eliminate the error case of an invalid string and make the below assert unnecessary

**Review by azuwis — COMMENTED** [review]
How about this?

**Comment by emilazy** [comment]
Implemented a variant of @azuwis’s suggestion. I prefer not using a boolean here and it seems pretty compact now.

**Review by paparodeo — APPROVED** [review]

**Comment by emilazy** [comment]
Going to take the thumbs up as the closest thing we’ll get to Ninja maintainer approval here :)

**Comment by emilazy** [comment]
Result of `nixpkgs-review pr 354367` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)                                                          
<details>                                                                       
  <summary>1 package built:</summary>                                           
  <ul>                                                                          
    <li>ninja_1_11</li>                                                         
  </ul>                                                                         
</details>

**Comment by AndersonTorres** [comment]
I want to study that patch from in memoriam @amjoseph that enables multiple versions. IIRC it populates passthru.versions with the various derivations.

**Comment by ncfavier** [comment]
This change unfortunately breaks `nix-prefetch ninja`, and as a consequence also `nix-prefetch invidious` and probably lots of other packages. I don't know why. See https://github.com/NixOS/nixpkgs/pull/355086#issuecomment-2468959055

```console
$ nix-prefetch -I nixpkgs=. ninja
error:
       … while evaluating attribute 'log'

       … while calling the 'head' builtin
         at /home/n/git/nixpkgs/lib/attrsets.nix:1574:11:
         1573|         || pred here (elemAt values 1) (head values) then
         1574|           head values
             |           ^
         1575|         else

       … while evaluating the attribute 'name'
         at /home/n/git/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:336:7:
          335|     // (optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          336|       name =
             |       ^
          337|         let

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: attribute 'rev' missing
       at /home/n/git/nixpkgs/pkgs/by-name/ni/ninja/package.nix:19:34:
           18|   pname = "ninja";
           19|   version = lib.removePrefix "v" finalAttrs.src.rev;
             |                                  ^
           20|
```

**Comment by emilazy** [comment]
I suspect `nix-prefetch` overrides `fetchFromGitHub` in a way that doesn't expose `rev` on the result, breaking the interface. That’s a `nix-prefetch` bug that should probably be fixed, although it would be possible to work around it in this package with a little contortion (to keep supporting `nix-update`, which is fusys in its own way).

**Comment by JohnRTitor** [comment]
I think the issue here is that the version depends on the rev, unlike most of the nixpkgs where it does not.

**Pull request merged**
